### PR TITLE
Editor Welcome: Add instructional graphics for mobile UI

### LINKED ISSUE #55826: Editor Welcome: instructional graphics for mobile UI
### What

When displaying the Editor Welcome tour on mobile, we should show instructional graphics relevant to the mobile screen resolution. 

Some of the slides included in the tour are generic (eg: block, block toolbar, drag & drop) but some of them show the editor UI ( p1629901900075300/1629899413.074600-slack-C029WSAQMT9 ) and probably those would need a mobile version.

Latest update on Slack: p1630298089013600/1629899413.074600-slack-C029WSAQMT9

### Why

Our users would benefit most from guidance adapted to the screen they are using.

### A/C

- [x] GIFs with the mobile UI @donalirl 
- [ ] Final graphics with the GIFs on grey background, similar to the desktop ones @poligilad-auto 
- [ ] (dev) Implement new graphics and copy to be rendered when the Welcome Tour is displayed on mobile

### PR DESCRIPTION
**Pull request by StefanNieuwenhuis**
### Changes proposed in this Pull Request

* Add mobile assets to Welcome Tour
 
#### Technical Changes
* Make use of `<picture/>` element to support different graphics on different viewports.
* Update `imgSrc` structure from string to map in order to support different graphics on different viewports.
* Update styling to maintain proportions of graphics

#### Preparations 
1. Make sure that you're sandboxed
2. Apply D67026-code w/ `arc patch-image` (see PCYsg-h9w-p2)
3. Make sure that`s0.wp.com` is sandboxed. 
4. Check out this branch and `yarn && yarn dev --sync`. 

#### Testing instructions
1. Navigate to one of your sandboxed sites and clear the browser's local storage.
2. Navigate to the editor
3. The welcome tour should pop up. If not, launch it from the _more settings_ menu.
4. Validate that:
    - [ ] All animations work as expected (see screencap below).
    - [ ] The animations for desktop are visible.
5. Switch to a mobile viewport
6. Repeat steps 1-3 and validate that:
    - [ ] All animations work as expected (see screencap below).
    - [ ] The animations for mobile are visible. 

### Welcome Tour on desktop screencap
https://user-images.githubusercontent.com/12104589/133770360-8adf4aad-bac1-495c-8474-a9b39431d2a0.mov

### Welcome Tour on mobile screencap
https://user-images.githubusercontent.com/12104589/133770349-98cbdbfc-6456-49f4-8785-6cc3cc06e593.mov

Fixes #55826

### CODE REVIEW

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (StefanNieuwenhuis) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: An alternative approach is to use `` and define the `srcset` attribute. The upside is that we don't have to introduce CSS overrides and/or request component changes. 

The downside is that you have to provide a _width_ or _density descriptor_ for each image (see [MDN Docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/picture#attr-srcset)). The width descriptor has to be the exact size of an image, which, IMHO, is not a very user-friendly solution. 

Besides that, the `<picture/>` is meant to provide **different** images for different viewports in a sense that the content is completely different. Using `srcset` is for using **same** images w/ different resolutions.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js
Comment (StefanNieuwenhuis) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js]: This is a structure I came up with to enable differing between desktop and mobile images. 
I used a map since it has O(n) complexity and it's descriptive as well. 

An alternative approach is to create a more extensive `media` mapping to support other formats (e.g. video) too but I thought of it as a follow-up (maybe in combination with other improvements to the tour content object).

Any other suggestions are welcome, though!

Comment (chriskmnds) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js]: Some sort of wrapper to encapsulate a little may not be a bad idea here. Not sure how I feel about having all these constants this way. 

Maybe something along the lines:

```
const mediaSources = {
    addBlock: {
        mobile: ...,
        desktop: ...,
    },
    allBlocks: ...,
    ...,
};
```

May even wrap that in a function `getMediaSources` and move `CDNPREFIX` there too. I guess tiny things to add a minimal structure as it feels like a quick win.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: NIT. It would be nice if we could make the media query or even `600px` more visible to the app.

Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: It looks like if we include the MIME `type`, it saves a round trip to query for it (just reading the specs below). If that's the case and we expect the type to be an image then might be worth optimizing?

> If the type attribute isn't specified, the media's type is retrieved from the server and checked to see if the user agent can handle it; if it can't be rendered, the next <source> is checked. If the type attribute is specified, it's compared against the types the user agent can present, and if it's not recognized, the server doesn't even get queried; instead, the next <source> element is checked at once.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source#attr-type

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/style-tour.scss
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/style-tour.scss:207]: Can we make this a little more clear so it doesn't linger this way? No idea if there's a plan to updated card-styles. Should we link to an issue or something? 🤔

**Review by chriskmnds — COMMENTED** [review]
Just a few comments reading through the code. Haven't tested anything. Nice use of the `picture` & `source` elements.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: I was not very happy with this solution either, since it feels quite hacky to hardcode the media queries. So, I hope there is some method/lib/util that serves the viewports.

This is something that we need with preloading images as well.
https://github.com/Automattic/wp-calypso/blob/2a012af405964741086cab1818eddc09b17e40eb/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js#L36

https://github.com/Automattic/wp-calypso/blob/2a012af405964741086cab1818eddc09b17e40eb/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js#L129 

We have to retrieve the viewport to determine which image to preload.

Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: Great find @chriskmnds! I added it in 651e469f60968dd0e355f30a23ae1f946a57981b.
Although it's unnecessary for now to introduce a MIME-type for desktop assets, I wanted to keep the data structure for both mobile & desktop consistent, hence I added it for both.

**Comment by StefanNieuwenhuis** [comment]
This PR breaks preloading images for the Welcome Tour (see https://github.com/Automattic/wp-calypso/pull/56347#discussion_r712872855), hence I put it on hold.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: ### Prefetching images
In 15e50c7abd1724e233c4608a49f78c9f7057c45f, I use the `isDestkop()` function from `@automattic/viewport` to determine the device and prefetch the right assets.

### Media query in picture source
`@automattic/viewport` contains a **private** method `getMediaQueryList`. It contains a list of all **valid** breakpoints in Calypso and converts it to a media query. Exactly what we need to populate the `media` property of our mobile assets in the Welcome Tour! Hence I exposed it and implemented it in 9f13f972b0b13c2701cce0bcd2206dfb6c4aa288. 

I'm not totally sure if this is the proper way to go, and you or maybe @razvanpapadopol have alternative approaches?

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js]: I totally agree with you, and I addressed it in 68325d3a9251a6a8584715d818008918543829ba

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: I think we don't need optional chaining here. `imgSrc.mobile` should be bound/safe from the outer condition?

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js]: I think this is not necessary either. `tourAssets[ undefined ]` will return `undefined`

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: NIT. We could add a tiny abstraction here if you like:

```
getImageSrcForView( card ) { 
    return ! isDesktop() && card?.imgSrc?.mobile ? card.imgSrc.mobile.src : card?.imgSrc?.desktop?.src
}
```

and can use it above for lines 37-38, changing to:

```
const card = getTourContent( localeSlug )[ 0 ];
new window.Image().src = getImageSrcForView( card );
```

**Review by chriskmnds — COMMENTED** [review]
Looking good Stefan. Nice follow-ups. 🚀 

Just some tiny comments while testing. Logic tests out for me, just the images not loading from the patch on my end 🤷

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js]: > I think this is not necessary either. `tourAssets[ undefined ]` will return `undefined`

Yep, that's correct but the function without the check will be significantly slower when called without arguments.

Screenshot 2021-09-24 at 11 24 31

Link to the benchmark test: https://jsben.ch/rbKVv

On the other hand, calling the function with arguments doesn't make much difference when it comes to speed. Without the check, the function is slightly faster.

Screenshot 2021-09-24 at 11 27 20

Link to the benchmark test: https://jsben.ch/D2xIW

Since we (currently) always call it with arguments (the function is only used in `getTourContent()`, where arguments are always provided), it makes sense to remove the check, which I did in af32ab174a79650981b04eb7c1154433eb36c49c

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: Great idea! I've implemented it in d7ff32c972b8aa4d4f41a363edd86dedb43248ca

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/types.ts
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/types.ts]: Since we don't use this anywhere, I'd avoid introducing it just yet. Also, we may decide to rethink the index signature once tour contents are migrated to TS - might be the case a closed-form or mapping would be more appropriate.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (chriskmnds) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: Very nice. Great to see more types sliding in.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js
Comment (chriskmnds) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-content.js]: Very nice. Always appreciate a good thought on optimization. As long as there's a pragmatic focus - to not feel premature and unnecessary.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: Technically speaking `getMediaQueryList` may return `undefined`. I guess we could guard irrespective if we want to be extra cautious. 🤷

**Review by chriskmnds — DISMISSED** [review]
Looks good. I don't have any more input here. Approving to get things rolling (if you are waiting for approval). Just need to sort out the images loading properly.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: Yep, you're right! The outer condition renders the need for optional chaining obsolete, so I fixed it in https://github.com/Automattic/wp-calypso/commit/af32ab174a79650981b04eb7c1154433eb36c49c.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js]: Thanks, @chriskmnds! I added a guard to be catch errors when non-supported breakpoints are used.

**Comment by StefanNieuwenhuis** [comment]
> Looks good. I don't have any more input here. Approving to get things rolling (if you are waiting for approval). Just need to sort out the images loading properly.

Thanks for reviewing and your insights, @chriskmnds 🙏 
I'll solve the merge conflicts, and publish once I figured out the image loading issue.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts]: Let's rethink this part and consider using [@automattic/viewport-react](https://www.npmjs.com/package/@automattic/viewport-react) instead. Might mean a component that finds the correct image based on viewport changes. 

Related thing happening in #58009

CC @escapemanuele @razvanpapadopol

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: > Preload first card image (others preloaded after open state confirmed)

Do we no longer want this optimization?

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: I feel we may be introducing a bit of an anti-pattern with calling JSX components in this fashion 🤔

Sorry if I implied something of this nature in my comment. Have we considered implementing a hook for this, something like `useImagePrefetcher`, and do the looping there maybe? So a signature of the form `( assets: TourAsset[] ) => void`, and see if also usable for the optimisation we are removing [above](https://github.com/Automattic/wp-calypso/blob/trunk/apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js#L31)?

Thanks for looking into this.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: Hmm, not sure what happened here (probably some rebasing stuff) but I reverted the changes.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js]: > I feel we may be introducing a bit of an anti-pattern with calling JSX components in this fashion 🤔

I agree. JSX should be used to render components and not to prefetch images (although, IMHO, it's quite cool that it's even possible!), hence a hook is a better solution. I've changed it in the latest commit.

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts]: Why not keep the signature `( steps: Steps ) => void` and always pass in an array? I don't think we need the extra complexity with dealing with both singular and multiple steps, and we can also keep the semantics sane ("steps" being plural).

File: apps/editing-toolkit/tsconfig.json
Comment (scinos) [inline apps/editing-toolkit/tsconfig.json]: I think you mean `../../packages/viewport-react`.

But it won't work, as `@automattic/viewport-react` is a JavaScript package that doesn't export types.

They have been declared in `packages/tour-kit/src/types-patch.d.ts`, but I believe we should be able to move them to `packages/viewport-react`. I'll try that.

Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/tsconfig.json]: Yes, it must be `@automattic/viewport-react`. We've changed that in one of the more recent commits but we forgot to update the refs.

Comment (escapemanuele) [reply apps/editing-toolkit/tsconfig.json]: Thanks @scinos, it would be great if types were in `packages/viewport-react`.

Comment (scinos) [reply apps/editing-toolkit/tsconfig.json]: I've pushed a commit with moving the types to `packages/viewport-react` and it seems to work!

At least it now fails somewhere else:

image

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts
Comment (chriskmnds) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts]: I think it would be a nice cleanup to also create a `hooks` folder and keep hooks contained there. So refactor this to `hooks/use-prefetch-tour-assets.ts` instead of a general-purpose utils file where anything goes.. 😬

Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts]: Done in cedcb2c

File: apps/editing-toolkit/tsconfig.json
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/tsconfig.json]: Thanks so much, @scinos! I fixed the other TS errors and all checks are passing 👍

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts
Comment (StefanNieuwenhuis) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/utils.ts]: Thanks for addressing this. 

In `apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js:30`, we pass a singular argument to the hook, hence we have to add additional logic to diverge between singular and multiple arguments. 

```js
// Preload first card image (others preloaded after open state confirmed)
new window.Image().src = usePrefetchTourAssets( getTourSteps( localeSlug )[ 0 ] );
```

It feels wrong to wrap a singular argument in an array just to satisfy a hook's signature but doing it makes the hook:
1. More comprehensible.
2. More performant.
3. Easier to reuse.

Therefore I refactored the code in fba3324.

**Review by chriskmnds — APPROVED** [review]
Looking good. :shipit: if it tests out

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js
Comment (arthur791004) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-launch.js:31]: Hmmm...I think `usePrefetchTourAssets` returns `undefined` and it would cause the page to load the undefined assets.

I think we only need the following one
```js
usePrefetchTourAssets( [ getTourSteps( localeSlug )[ 0 ] ] );
```

File: apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js
Comment (razvanpapadopol) [inline apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js:68]: Instead of hardcoding `'<480px'`, can we import the `MOBILE_BREAKPOINT` const from the same package? I think this should help our future selves understand a bit better the use of `@automattic/viewport` here.

Comment (escapemanuele) [reply apps/editing-toolkit/editing-toolkit-plugin/wpcom-block-editor-nux/src/welcome-tour/tour-card.js:68]: Added in https://github.com/Automattic/wp-calypso/pull/59052

**Pull request merged**
